drivers/pms5003/pms5003.c

```diff
@@ -0,0 +1,468 @@
+    mutex_lock(&dev->internal.busy);
+    DEBUG_PUTS("[pms5003]: busy");
+    ztimer_set(ZTIMER_USEC, &dev->internal.busy_timer, time);
+}
+
+int _write_cmd(pms5003_t *dev, pms5003_cmd_t cmd, uint16_t data)
+{
+    uint8_t buf[PMS5003_CMD_FRAME_LEN];
+
+    buf[0] = PMS5003_START_BYTE_1;
+    buf[1] = PMS5003_START_BYTE_2;
+    buf[2] = cmd;
+    buf[3] = (uint8_t)(data >> 8);
+    buf[4] = (uint8_t)data;
+    /* calculate crc */
+    uint16_t crc = 0;
+    for (uint8_t i = 0; i < 5; i++) {
+        crc += buf[i];
+    }
+    buf[5] = (uint8_t)(crc >> 8);
+    buf[6] = (uint8_t)crc;
+    /* set busy to avoid sendinf commands before PMS5003_CMD_BUSY_TIME of
+       a previously issued command */
+    _set_busy(dev, PMS5003_CMD_BUSY_TIME);
+    /* write cmd */
+    uart_write(dev->params.uart, buf, sizeof(buf));
+    /* wait for response or timeout */
+    return _wait_for_resp(dev);
+}
+
+void _handle_received(pms5003_t *dev)
+{
+    /* parse data */
+    pms5003_data_t data;
+    for (uint8_t i = 0; i < sizeof(data.buffer) / 2; i++) {
+        data.buffer[i] = byteorder_ntohs(dev->internal.parser.buf.be[PMS5003_PARSER_DATA_IDX + i]);
+    }
+    for (pms5003_callbacks_t *i = dev->cbs; i != NULL; i = i->next) {
+        if (i->cb) {
+            i->cb(&data, i->arg);
+        }
+    }
+}
+
+int _parse_frame_content(pms5003_t *dev)
+{
+    int ret = -1;
+
+    switch (dev->state) {
+    case PMS5003_STATE_MODE_ACTIVE_REQUEST:
+        if (dev->internal.parser.buf.be[PMS5003_PARSER_DATA_IDX].u16 == PMS5003_CMD_ACTIVE_RESPONSE) {
+            ret = 0;
+            dev->state = PMS5003_STATE_READ_REQUEST;
+            mutex_unlock(&dev->internal.timeout);
+        }
+        break;
+    case PMS5003_STATE_MODE_PASSIVE_REQUEST:
+        if (dev->internal.parser.buf.be[PMS5003_PARSER_DATA_IDX].u16 == PMS5003_CMD_PASSIVE_RESPONSE) {
+            ret = 0;
+            dev->state = PMS5003_STATE_IDLE;
+            mutex_unlock(&dev->internal.timeout);
+        }
+        break;
+    case PMS5003_STATE_SLEEP_REQUEST:
+        if (dev->internal.parser.buf.be[PMS5003_PARSER_DATA_IDX].u16 == PMS5003_CMD_SLEEP_RESPONSE) {
+            ret = 0;
+            dev->state = PMS5003_STATE_SLEEP;
+            mutex_unlock(&dev->internal.timeout);
+        }
+        break;
+    case PMS5003_STATE_READ_REQUEST:
+        if (dev->mode == PMS5003_PASSIVE_MODE) {
+            dev->state = PMS5003_STATE_IDLE;
+        }
+        ret = 0;
+        mutex_unlock(&dev->internal.timeout);
+        if (dev->cbs != NULL) {
+            _handle_received(dev);
+        }
+    default:
+        break;
+    }
+    return ret;
+}
+
+int _verify_frame(pms5003_t *dev)
+{
+    uint16_t crc = 0;
+    uint16_t frame_crc = 0;
+    int8_t len;
+
+    if (dev->internal.parser.idx == PMS5003_RESP_FRAME_LEN - 1) {
+        len = PMS5003_RESP_FRAME_LEN - PMS5003_CRC_LEN;
+        frame_crc = byteorder_ntohs(dev->internal.parser.buf.be[PMS5003_PARSER_RESP_CRC_IDX]);
```

Comment:
Note that there is `uint16_t byteorder_bebuftohs(const uint8_t *buf)`. IMO the code would be easier to read if you would ditch this the hole parser struct and just add an `uint8_t buf[]` into the device descriptor, and access big-ending 16 bit numbers via byteorder_bebuftohs()